addons/mail/models/mail_thread.py

```diff
@@ -2220,7 +2220,9 @@ def _notify_record_by_email(self, message, recipients_data, msg_vals=False,
 
         model = msg_vals.get('model') if msg_vals else message.model
         model_name = model_description or (self.with_lang().env['ir.model']._get(model).display_name if model else False) # one query for display name
-        recipients_groups_data = self._notify_classify_recipients(partners_data, model_name)
+        res_id = msg_vals.get('res_id') if msg_vals else message.res_id
+        access_link = self._notify_get_action_link('view', res_id=res_id, model=model)
+        recipients_groups_data = self._notify_classify_recipients(partners_data, model_name, access_link)
```

Comment:
My initial suggestion to avoid to change method api was to move _notify_get_action_link inside _notify_prepare_template_context, and instead of chosing yourself res_id and res_model just use **msg_vals as kwargs.
Since group_values will update template_value, it should work I think.